crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2615,7 +2615,18 @@ impl<'db> TypeInferenceBuilder<'db> {
                 if !dunder_getitem_method.is_unbound() {
                     return dunder_getitem_method
                         .call(self.db, &[slice_ty])
-                        .unwrap_with_diagnostic(self.db, value.as_ref().into(), self);
+                        .return_ty_result(self.db, value.as_ref().into(), self)
+                        .unwrap_or_else(|err| {
```

Comment:
This kind of chaining also lets us avoid monomorphizing `unwrap_with_diagnostic`.

Reply:
The new pattern looks great to me, but I'm not sure I understand the comment. `unwrap_with_diagnostic` is not generic (except over a lifetime), so what does it mean to monomorphize it?

Reply:
Ah sorry, that was a confusing comment out-of-context. I initially considered having `unwrap_with_diagnostic` accept a callback function to format the diagnostics, which would've _then_ made it generic on that callback. By returning result, I avoided making it generic :)